File: androidApp/src/main/kotlin/com/karim/foodrun/GroupHomeContent.kt
package com.karim.foodrun

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.material3.TextButton
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import com.karim.foodrun.shared.FoodRunText
import com.karim.foodrun.shared.orders.*

@Composable
internal fun GroupHomeContent(state: GroupState, controller: GroupController) {
    Column(verticalArrangement = Arrangement.spacedBy(FoodSpacing.Page)) {
        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(FoodSpacing.Small)) {
            Text(FoodRunText.brand, style = FoodType.Brand, color = FoodColors.Ink, modifier = Modifier.weight(1f))
            GroupLanguagePicker(state, controller)
        }
        Column(verticalArrangement = Arrangement.spacedBy(FoodSpacing.XSmall)) {
            Text(GroupText.localized(GroupText.homeTitle, state.rtl), style = FoodType.Hero, color = FoodColors.Ink, modifier = Modifier.semantics { heading() })
            Text(GroupText.localized(GroupText.homeSubtitle, state.rtl), style = FoodType.Body, color = FoodColors.Muted)
        }
        FoodCard(fill = FoodColors.Hero, radius = FoodSpacing.Page) {
            Column(Modifier.padding(FoodSpacing.Page), verticalArrangement = Arrangement.spacedBy(FoodSpacing.Large)) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(GroupText.localized(GroupText.groupEyebrow, state.rtl), style = FoodType.RoundedCaption, color = FoodColors.OnHero, modifier = Modifier.weight(1f))
                    Icon(Icons.Default.Person, null, tint = FoodColors.OnHero)
                }
                Text(GroupText.localized(GroupText.groupTitle, state.rtl), style = FoodType.DialogTitle, color = FoodColors.White)
                Text(GroupText.localized(GroupText.groupDescription, state.rtl), style = FoodType.Body, color = FoodColors.OnHero)
                state.buttons.filter { it.action in listOf(GroupAction.CREATE, GroupAction.JOIN) }.forEach {
                    GroupActionButton(it, state.busy, controller)
                }
            }
        }
        state.cards.filter { it.id.startsWith("invitation:") }.forEach { GroupCardContent(it, state.busy, controller) }
        val rooms = state.cards.filter { it.id.startsWith("session:") }
        if (rooms.isNotEmpty()) {
            Column(verticalArrangement = Arrangement.spacedBy(FoodSpacing.Medium)) {
                GroupSectionHeading(GroupText.localized(GroupText.savedRooms, state.rtl), rooms.size)
                rooms.forEach { GroupCardContent(it, state.busy, controller) }
            }
        }
        val dashboard = state.cards.filter { it.id.startsWith("dashboard:") }
        if (dashboard.isNotEmpty()) {
            Column(verticalArrangement = Arrangement.spacedBy(FoodSpacing.Medium)) {
                GroupSectionHeading(if (state.rtl) "المحفظة والطلبات" else "Wallet & orders")
                dashboard.forEach { GroupCardContent(it, state.busy, controller) }
            }
        }
        Column(verticalArrangement = Arrangement.spacedBy(FoodSpacing.Medium)) {
            GroupSectionHeading(GroupText.localized(GroupText.explore, state.rtl))
            state.buttons.filter { it.action in listOf(GroupAction.QUICK_SPIN, GroupAction.OPEN_LIBRARY) }.forEach { button ->
                FoodCard(bordered = true, modifier = Modifier.testTag("action:${button.action.name}:${button.value}")) {
                    Row(Modifier.fillMaxWidth().clickable(enabled = !state.busy && button.enabled, role = Role.Button) {
                        controller.dispatch(button.action, button.value)
                    }.padding(FoodSpacing.Large), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(FoodSpacing.Content)) {
                        GroupIcon(groupActionIcon(button), accented = button.action == GroupAction.QUICK_SPIN)
                        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(FoodSpacing.XXSmall)) {
                            Text(button.title, style = FoodType.Person, color = FoodColors.Ink)
                            Text(if (button.action == GroupAction.QUICK_SPIN) GroupText.localized(GroupText.quickDescription, state.rtl) else GroupText.localized(GroupText.libraryDescription, state.rtl), style = FoodType.Caption, color = FoodColors.Muted)
                        }
                        Icon(Icons.AutoMirrored.Filled.ArrowForward, null, tint = FoodColors.Muted, modifier = Modifier.size(FoodSize.IconSmall))
                    }
                }
            }
        }
        state.buttons.filter { it.action !in listOf(GroupAction.CREATE, GroupAction.JOIN, GroupAction.QUICK_SPIN, GroupAction.OPEN_LIBRARY, GroupAction.SET_LANGUAGE) }.forEach {
            GroupActionButton(it, state.busy, controller)
        }
        state.cards.firstOrNull { it.id == "about" }?.let {
            Row(horizontalArrangement = Arrangement.spacedBy(FoodSpacing.Small)) {
                Icon(Icons.Default.Check, null, tint = FoodColors.Muted, modifier = Modifier.size(FoodSize.IconSmall))
                Text(it.detail, style = FoodType.Caption, color = FoodColors.Muted)
            }
        }
    }
}

@Composable
internal fun GroupIcon(icon: ImageVector, accented: Boolean = false) {
    Box(Modifier.size(FoodSize.TouchTarget).background(if (accented) FoodColors.AccentWash else FoodColors.SuccessWash, RoundedCornerShape(FoodRadius.Avatar)), contentAlignment = Alignment.Center) {
        Icon(icon, null, tint = if (accented) FoodColors.Orange else FoodColors.Success, modifier = Modifier.size(FoodSize.IconMedium))
    }
}

@Composable
internal fun GroupSectionHeading(title: String, count: Int? = null) {
    Row(Modifier.fillMaxWidth().padding(top = FoodSpacing.XSmall), verticalAlignment = Alignment.CenterVertically) {
        Text(title, style = FoodType.Person, color = FoodColors.Ink, modifier = Modifier.weight(1f).semantics { heading() })
        count?.let { Text(it.toString(), style = FoodType.Status, color = FoodColors.Muted) }
    }
}

internal fun groupActionIcon(button: GroupButton): ImageVector = when {
    button.destructive -> Icons.Default.Delete
    else -> when (button.action) {
        GroupAction.CREATE, GroupAction.CREATE_ROOM, GroupAction.NEW_RESTAURANT, GroupAction.ADD_MENU_ITEM, GroupAction.ADD_CART_ITEM -> Icons.Default.Add
        GroupAction.JOIN, GroupAction.JOIN_ROOM, GroupAction.RESUME -> Icons.Default.Person
        GroupAction.USE_INTERNET -> Icons.Default.Share
        GroupAction.QUICK_SPIN, GroupAction.PREPARE_SPIN, GroupAction.REFRESH, GroupAction.RETRY -> Icons.Default.Refresh
        GroupAction.OPEN_LIBRARY, GroupAction.OPEN_ITEM, GroupAction.OPEN_RECEIPTS -> Icons.Default.Menu
        GroupAction.OPEN_HISTORY -> Icons.Default.DateRange
        GroupAction.FAVORITE_ORDER, GroupAction.REMOVE_FAVORITE_ORDER -> Icons.Default.Favorite
        GroupAction.REUSE_ORDER -> Icons.Default.ShoppingCart
        GroupAction.SHARE_ROOM, GroupAction.SHARE_RECEIPT, GroupAction.EXPORT_MENU, GroupAction.SHARE_ORDER_WHATSAPP -> Icons.Default.Share
        GroupAction.READY, GroupAction.CONFIRM_QUOTE, GroupAction.SAVE_RESTAURANT, GroupAction.ACCEPT_DUTY,
        GroupAction.SELECT_TAX_TREATMENT -> Icons.Default.Check
        GroupAction.EDIT_RESTAURANT, GroupAction.EDIT_ROOM_RESTAURANT -> Icons.Default.Edit
        else -> Icons.AutoMirrored.Filled.ArrowForward
    }
}


@Composable
internal fun GroupLanguagePicker(state: GroupState, controller: GroupController) {
    var expanded by remember { mutableStateOf(false) }
    Box {
        TextButton(onClick = { expanded = true }, enabled = !state.busy,
            modifier = Modifier.heightIn(min = FoodSize.TouchTarget).testTag("languagePicker")
                .semantics { contentDescription = "Language / اللغة" }) {
            androidx.compose.foundation.Canvas(Modifier.size(FoodSize.IconSmall)) {
                val stroke = androidx.compose.ui.graphics.drawscope.Stroke(width = size.width / 14)
                drawCircle(FoodColors.Orange, radius = size.width * .44f, style = stroke)
                drawOval(FoodColors.Orange, topLeft = androidx.compose.ui.geometry.Offset(size.width * .28f, size.height * .06f), size = androidx.compose.ui.geometry.Size(size.width * .44f, size.height * .88f), style = stroke)
                drawLine(FoodColors.Orange, androidx.compose.ui.geometry.Offset(size.width * .06f, size.height / 2), androidx.compose.ui.geometry.Offset(size.width * .94f, size.height / 2), strokeWidth = stroke.width)
            }
            Text(if(state.rtl) "العربية" else "English", modifier = Modifier.padding(horizontal = FoodSpacing.XSmall))
            Icon(Icons.Default.KeyboardArrowDown, null, modifier = Modifier.size(FoodSize.IconSmall))
        }
        DropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            listOf("en" to "English", "ar" to "العربية").forEach { (code, label) ->
                DropdownMenuItem(text = { Text(label) }, onClick = {
                    expanded = false; controller.dispatch(GroupAction.SET_LANGUAGE, code)
                }, trailingIcon = { if ((code == "ar") == state.rtl) Icon(Icons.Default.Check, null) })
            }
        }
    }
}

File: FoodRun/GroupHomeContent.swift
import SwiftUI
import FoodRunShared

struct GroupHomeContent: View {
    let state: GroupState
    let dispatch: (GroupAction, String) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: FoodSpacing.s24) {
            HStack(spacing: FoodSpacing.s10) {
                Text(FoodStrings.text.brand).font(FoodTypography.brand).tracking(FoodSpacing.s2)
                Spacer()
                GroupLanguagePicker(state: state, dispatch: dispatch)
            }
            if let signIn = state.cards.first(where: { $0.id == "sign-in" }) {
                GroupCardContent(card: signIn, busy: state.busy, dispatch: dispatch)
                ForEach(state.buttons.filter { $0.action != .setLanguage }, id: \.renderID) { button in
                    GroupActionContent(button: button, busy: state.busy, dispatch: dispatch)
                }
            } else {
            VStack(alignment: .leading, spacing: FoodSpacing.s8) {
                Text(GroupText.shared.localized(value: GroupText.shared.homeTitle, rtl: state.rtl)).font(FoodTypography.hero).tracking(-1)
                    .fixedSize(horizontal: false, vertical: true).accessibilityAddTraits(.isHeader)
                Text(GroupText.shared.localized(value: GroupText.shared.homeSubtitle, rtl: state.rtl)).font(FoodTypography.subtitle).foregroundStyle(FoodTheme.muted)
            }
            VStack(alignment: .leading, spacing: FoodSpacing.s16) {
                HStack {
                    Text(GroupText.shared.localized(value: GroupText.shared.groupEyebrow, rtl: state.rtl)).font(FoodTypography.eyebrow).tracking(FoodSpacing.s1)
                    Spacer(minLength: FoodSpacing.s8)
                    Image(systemName: "person.3.fill").font(FoodTypography.brandIcon).accessibilityHidden(true)
                }.foregroundStyle(FoodTheme.onHero)
                Text(GroupText.shared.localized(value: GroupText.shared.groupTitle, rtl: state.rtl)).font(FoodTypography.formTitle).foregroundStyle(FoodTheme.white)
                Text(GroupText.shared.localized(value: GroupText.shared.groupDescription, rtl: state.rtl)).font(FoodTypography.subtitle).foregroundStyle(FoodTheme.onHero)
                ForEach(state.buttons.filter { $0.action == .create || $0.action == .join }, id: \.renderID) { button in
                    GroupActionContent(button: button, busy: state.busy, dispatch: dispatch)
                }
            }
            .padding(FoodSpacing.s24)
            .background(FoodTheme.hero, in: RoundedRectangle(cornerRadius: FoodRadius.group))

            ForEach(state.cards.filter { $0.id.hasPrefix("invitation:") }, id: \.renderID) { card in GroupCardContent(card: card, busy: state.busy, dispatch: dispatch) }
            let rooms = state.cards.filter { $0.id.hasPrefix("session:") }
            if !rooms.isEmpty {
                VStack(alignment: .leading, spacing: FoodSpacing.s12) {
                    GroupSectionHeading(title: GroupText.shared.localized(value: GroupText.shared.savedRooms, rtl: state.rtl), count: rooms.count)
                    ForEach(rooms, id: \.renderID) { card in GroupCardContent(card: card, busy: state.busy, dispatch: dispatch) }
                }
            }
            let dashboard = state.cards.filter { $0.id.hasPrefix("dashboard:") }
            if !dashboard.isEmpty {
                VStack(alignment: .leading, spacing: FoodSpacing.s12) {
                    GroupSectionHeading(title: state.rtl ? "المحفظة والطلبات" : "Wallet & orders")
                    ForEach(dashboard, id: \.renderID) { card in
                        GroupCardContent(card: card, busy: state.busy, dispatch: dispatch)
                    }
                }
            }
            VStack(alignment: .leading, spacing: FoodSpacing.s12) {
                GroupSectionHeading(title: GroupText.shared.localized(value: GroupText.shared.explore, rtl: state.rtl))
                ForEach(state.buttons.filter { $0.action == .quickSpin || $0.action == .openLibrary }, id: \.renderID) { button in
                    Button { dispatch(button.action, button.value) } label: {
                        HStack(spacing: FoodSpacing.s14) {
                            GroupIcon(symbol: button.symbol, accented: button.action == .quickSpin)
                            VStack(alignment: .leading, spacing: FoodSpacing.s5) {
                                Text(button.title).font(FoodTypography.bodyBold).foregroundStyle(FoodTheme.ink)
                                Text(button.action == .quickSpin ? GroupText.shared.localized(value: GroupText.shared.quickDescription, rtl: state.rtl) : GroupText.shared.localized(value: GroupText.shared.libraryDescription, rtl: state.rtl))
                                    .font(FoodTypography.footer).foregroundStyle(FoodTheme.muted)
                            }
                            Spacer(minLength: FoodSpacing.s0)
                            Image(systemName: "arrow.up.right").foregroundStyle(FoodTheme.muted).accessibilityHidden(true)
                        }.padding(FoodSpacing.s16).frame(maxWidth: .infinity, alignment: .leading).foodCard(showsBorder: true)
                    }
                    .buttonStyle(PressableStyle()).disabled(state.busy || !button.enabled)
                    .accessibilityIdentifier("action:\(button.action.name):\(button.value)")
                }
            }
            ForEach(state.buttons.filter { ![GroupAction.create, .join, .quickSpin, .openLibrary, .setLanguage].contains($0.action) }, id: \.renderID) { button in
                GroupActionContent(button: button, busy: state.busy, dispatch: dispatch)
            }
            if let about = state.cards.first(where: { $0.id == "about" }) {
                Label(about.detail, systemImage: "checkmark.shield")
                    .font(FoodTypography.footer).foregroundStyle(FoodTheme.muted)
                    .fixedSize(horizontal: false, vertical: true)
            }
            }
        }.foregroundStyle(FoodTheme.ink)
    }
}


struct GroupLanguagePicker: View {
    let state: GroupState
    let dispatch: (GroupAction, String) -> Void
    var body: some View {
        Menu {
            Picker("Language / اللغة", selection: Binding(get: { state.rtl ? "ar" : "en" }, set: { dispatch(.setLanguage, $0) })) {
                Text("English").tag("en")
                Text("العربية").tag("ar")
            }
        } label: {
            HStack(spacing: FoodSpacing.s8) {
                Image(systemName: "globe")
                Text(state.rtl ? "العربية" : "English")
                Image(systemName: "chevron.down").imageScale(.small)
            }
            .font(FoodTypography.setting)
            .padding(.horizontal, FoodSpacing.s12)
            .frame(minHeight: FoodSpacing.s44)
            .background(FoodTheme.white, in: Capsule())
        }
        .disabled(state.busy)
        .accessibilityLabel("Language / اللغة")
        .accessibilityValue(state.rtl ? "العربية" : "English")
        .accessibilityIdentifier("languagePicker")
    }
}
